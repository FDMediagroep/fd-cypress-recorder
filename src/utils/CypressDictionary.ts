export enum FdEventType {
    ATTRIBUTE_VALUE_CONTAINS = 'attribute-value-contains',
    ATTRIBUTE_VALUE_EQUALS = 'attribute-value-equals',
    ATTRIBUTE_VALUE_EXISTS = 'attribute-value-exists',
    CLEAR_COOKIES = 'clear-cookies',
    CLICK = 'click',
    CONTAINS_TEXT = 'contains-text',
    HOVER = 'hover',
    LOCATION = 'location',
    LOCATION_CONTAINS = 'location-contains',
    EXISTS = 'exists',
    TYPE = 'type',
    VIEWPORT_SIZE = 'viewport-size',
    VISIT = 'visit',
}

export interface FdEvent {
    type: FdEventType;
}

export interface FdAttributeValueEvent extends FdEvent {
    target: string; // CSS Selector
    name: string;
    value: string;
}

export interface FdAttributeExistsEvent extends FdEvent {
    target: string; // CSS Selector
    name: string;
}

export interface FdTypeEvent extends FdEvent {
    target: string; // CSS Selector
    value: string;
}

export interface FdClickEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdHoverEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdLocationEvent extends FdEvent {
    href: string;
}

export interface FdLocationContainsEvent extends FdEvent {
    value: string;
}

export interface FdVisitEvent extends FdEvent {
    href: string;
}

export interface FdTextContentEvent extends FdEvent {
    target: string; // CSS Selector
    value: string;
}

export interface FdViewportSizeEvent extends FdEvent {
    width: number;
    height: number;
}

export interface FdExistsEvent extends FdEvent {
    target: string; // CSS Selector
}

export type AllFdEvents = FdEvent
    | FdAttributeValueEvent
    | FdAttributeExistsEvent
    | FdClickEvent
    | FdHoverEvent
    | FdLocationEvent
    | FdLocationContainsEvent
    | FdVisitEvent
    | FdTextContentEvent
    | FdViewportSizeEvent
    | FdExistsEvent
    | FdTypeEvent;

export interface Template {
    name: string;
    description?: string;
    events: AllFdEvents[];
}

export const UNIQUE_SELECTOR_OPTIONS = {
    selectorTypes: ['Tag', 'NthChild']
};

export interface Options {
    basicAuth?: boolean;
}

export function getCode(event: AllFdEvents, options?: Options) {
    switch (event.type) {
        case FdEventType.CLEAR_COOKIES:
            return `cy.clearCookies();`;
        case FdEventType.CLICK:
            return `cy.get('${(event as FdClickEvent).target}').should('exist').click();`;
        case FdEventType.CONTAINS_TEXT:
            return `cy.get('${(event as FdTextContentEvent).target}').contains('${(event as FdTextContentEvent).value.replace(new RegExp("'", 'g'), "\\\'")}');`;
        case FdEventType.EXISTS:
            return `cy.get('${(event as FdExistsEvent).target}').should('exist');`;
        case FdEventType.HOVER:
            return `cy.get('${(event as FdHoverEvent).target}').trigger('mouseover');`;
        case FdEventType.LOCATION:
            return `cy.location('href', {timeout: 10000}).should('eq', '${(event as FdLocationEvent).href}');`;
        case FdEventType.LOCATION_CONTAINS:
            return `cy.location('href', {timeout: 10000}).should('contain', '${(event as FdLocationContainsEvent).value}');`;
        case FdEventType.VISIT:
            if (options && options.basicAuth) {
                return `cy.visit('${(event as FdLocationEvent).href}', {auth: {username: Cypress.env('BASIC_USER') || '', password: Cypress.env('BASIC_PASS') || ''}});`;
            } else {
                return `cy.visit('${(event as FdLocationEvent).href}');`;
            }
        case FdEventType.TYPE:
            return `cy.get('${(event as FdTypeEvent).target}').type('${(event as FdTypeEvent).value}');`;
        case FdEventType.VIEWPORT_SIZE:
            return `cy.viewport(${(event as FdViewportSizeEvent).width}, ${(event as FdViewportSizeEvent).height});`;
        case FdEventType.ATTRIBUTE_VALUE_EQUALS:
            return `cy.get('${(event as FdAttributeValueEvent).target}').then((el: any) => el.attr('${(event as FdAttributeValueEvent).name}')).then((attr: string) => expect(attr).to.eq('${(event as FdAttributeValueEvent).value.replace(new RegExp("'", 'g'), "\\\'")}'))`;
        case FdEventType.ATTRIBUTE_VALUE_CONTAINS:
            return `cy.get('${(event as FdAttributeValueEvent).target}').then((el: any) => el.attr('${(event as FdAttributeValueEvent).name}')).then((attr: string) => expect(attr).to.contain('${(event as FdAttributeValueEvent).value.replace(new RegExp("'", 'g'), "\\\'")}'))`;
        case FdEventType.ATTRIBUTE_VALUE_EXISTS:
            return `cy.get('${(event as FdAttributeExistsEvent).target}').should('have.attr', '${(event as FdAttributeExistsEvent).name}');`;
        default:
            return `// ${JSON.stringify(event)}`;
    }
}
