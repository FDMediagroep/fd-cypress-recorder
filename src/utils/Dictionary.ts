import { AllFdEvents,
    Options,
    FdEventType,
    FdClickEvent,
    FdTextContentEvent,
    FdCountEqualsEvent,
    FdExistsEvent,
    FdHoverEvent,
    FdLocationEvent,
    FdLocationContainsEvent,
    FdTypeEvent,
    FdViewportSizeEvent,
    FdAttributeValueEvent,
    FdAttributeExistsEvent
} from "./FdEvents";

/**
 * Converts event to corresponding code and returns the code.
 * In case there is no translation the `event` will be returned as string.
 *
 * Supported events:
 *
 * FdEventType.ATTRIBUTE_VALUE_CONTAINS: check if given element has given attribute with value containing given value
 *
 * FdEventType.ATTRIBUTE_VALUE_EQUALS: check if given element has given attribute with value equaling given value
 *
 * FdEventType.ATTRIBUTE_VALUE_EXISTS: check if given element has given attribute
 *
 * FdEventType.CLEAR_COOKIES: clear browser session cookies
 *
 * FdEventType.CLICK: click an element
 *
 * FdEventType.CONTAINS_TEXT: check if an element contains given text
 *
 * FdEventType.COUNT_EQUALS: check if an element type exists given number of times within parent
 *
 * FdEventType.COUNT_GREATER_THAN: check if an element type exists more than given number of times within parent
 *
 * FdEventType.COUNT_LESS_THAN: check if an element type exists less than given number of times within parent
 *
 * FdEventType.EXISTS: check if element exists
 *
 * FdEventType.HOVER: set the mouseover state on element
 *
 * FdEventType.LOCATION: check if current location addressbar is the same as given value
 *
 * FdEventType.LOCATION_CONTAINS: check if current location value in addressbar contains given value.
 *                                This is useful to bypass unpredictable values in the URL. E.g. session hashes.
 *
 * FdEventType.VISIT: navigate to the given URL
 *
 * FdEventType.TYPE: type a certain text into the focused element
 *
 * FdEventType.VIEWPORT_SIZE: set the viewport size
 *
 * @param event
 * @param options
 */
export function getCodeFromEvent(event: AllFdEvents, options?: Options): string {
    switch (event.type) {
        case FdEventType.ATTRIBUTE_VALUE_CONTAINS:
            return `cy.get('${(event as FdAttributeValueEvent).target}').then((el) => expect(el.attr('${(event as FdAttributeValueEvent).name}')).to.contain('${(event as FdAttributeValueEvent).value.replace(new RegExp("'", 'g'), "\\\'")}'));`;
        case FdEventType.ATTRIBUTE_VALUE_EQUALS:
            return `cy.get('${(event as FdAttributeValueEvent).target}').then((el) => expect(el.attr('${(event as FdAttributeValueEvent).name}')).to.eq('${(event as FdAttributeValueEvent).value.replace(new RegExp("'", 'g'), "\\\'")}'));`;
        case FdEventType.ATTRIBUTE_VALUE_EXISTS:
            return `cy.get('${(event as FdAttributeExistsEvent).target}').should('have.attr', '${(event as FdAttributeExistsEvent).name}');`;
        case FdEventType.CLEAR_COOKIES:
            return `cy.clearCookies();`;
        case FdEventType.CLICK:
            return `cy.get('${(event as FdClickEvent).target}').click();`;
        case FdEventType.CONTAINS_TEXT:
            return `cy.get('${(event as FdTextContentEvent).target}').contains('${(event as FdTextContentEvent).value.replace(new RegExp("'", 'g'), "\\\'")}');`;
        case FdEventType.COUNT_EQUALS:
            return `cy.get('${(event as FdCountEqualsEvent).target}').should('have.length', ${(event as FdCountEqualsEvent).value});`;
        case FdEventType.COUNT_GREATER_THAN:
            return `cy.get('${(event as FdCountEqualsEvent).target}').should('have.length.gt', ${(event as FdCountEqualsEvent).value});`;
        case FdEventType.COUNT_LESS_THAN:
            return `cy.get('${(event as FdCountEqualsEvent).target}').should('have.length.lt', ${(event as FdCountEqualsEvent).value});`;
        case FdEventType.EXISTS:
            return `cy.get('${(event as FdExistsEvent).target}').should('exist');`;
        case FdEventType.HOVER:
            return `cy.get('${(event as FdHoverEvent).target}').trigger('mouseover');`;
        case FdEventType.LOCATION:
            return `cy.location('href', {timeout: 10000}).should('eq', '${(event as FdLocationEvent).href}');`;
        case FdEventType.LOCATION_CONTAINS:
            return `cy.location('href', {timeout: 10000}).should('contain', '${(event as FdLocationContainsEvent).value}');`;
        case FdEventType.VISIT:
            let opt = '';
            if (options) {
                if (options.basicAuth) {
                    opt = `auth: {username: Cypress.env('BASIC_USER') || '', password: Cypress.env('BASIC_PASS') || ''}`;
                }
                if (options.headers && options.headers.length) {
                    let headers = '';
                    options.headers.forEach((h) => {
                        if (h.property !== '') {
                            const prop = h.property.replace(/\\/g, '\\\\').replace(new RegExp('"', 'g'), '\\"');
                            const value = h.value.replace(/\\/g, '\\\\').replace(new RegExp('"', 'g'), '\\"');
                            if (headers) {
                                headers += `, "${prop}": "${value}"`;
                            } else {
                                headers += `"${prop}": "${value}"`;
                            }
                        }
                    });
                    headers = `headers: {${headers}}`;
                    if (opt) {
                        opt = `${opt}, ${headers}`;
                    } else {
                        opt = headers;
                    }
                }
                if (opt) {
                    opt = `, {${opt}}`;
                }
            }
            return `cy.visit('${(event as FdLocationEvent).href}'${opt});`;
        case FdEventType.TYPE:
            return `cy.get('${(event as FdTypeEvent).target}').type('${(event as FdTypeEvent).value}');`;
        case FdEventType.VIEWPORT_SIZE:
            return `cy.viewport(${(event as FdViewportSizeEvent).width}, ${(event as FdViewportSizeEvent).height});`;
        default:
            return `// ${JSON.stringify(event)}`;
    }
}

export function getCode(suite: string, description: string, events: AllFdEvents[], options?: Options) {
    const code: string[] = [
`/**
  * Code generated with Fd Cypress Recorder.
  * https://github.com/FDMediagroep/fd-cypress-recorder
  */

/// <reference types="Cypress" />
describe('${suite}', () => {
  beforeEach(() => {
    cy.clearCookies();
  });

  it('${description}', () => {
`];
    events.forEach((event) => {
        code.push(`    ${getCodeFromEvent(event, options)}\r\n`);
    });
    code.push('  });\r\n');
    code.push('});\r\n');

    return code;
}
