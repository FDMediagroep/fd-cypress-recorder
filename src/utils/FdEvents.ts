export interface Header {
    property: string;
    value: string;
}

export enum FdEventType {
    ATTRIBUTE_VALUE_CONTAINS = 'attribute-value-contains',
    ATTRIBUTE_VALUE_EQUALS = 'attribute-value-equals',
    ATTRIBUTE_VALUE_EXISTS = 'attribute-value-exists',
    CLEAR_COOKIES = 'clear-cookies',
    CLICK = 'click',
    CONTAINS_TEXT = 'contains-text',
    COUNT_EQUALS = 'count-equals',
    COUNT_GREATER_THAN = 'count-greater-than',
    COUNT_LESS_THAN = 'count-less-than',
    HOVER = 'hover',
    LOCATION = 'location',
    LOCATION_CONTAINS = 'location-contains',
    EXISTS = 'exists',
    /**
     * Type text
     */
    TYPE = 'type',
    VIEWPORT_SIZE = 'viewport-size',
    VISIT = 'visit',
}

export interface FdEvent {
    type: FdEventType;
}

export interface FdAttributeValueEvent extends FdEvent {
    target: string; // CSS Selector
    name: string;
    value: string;
}

export interface FdAttributeExistsEvent extends FdEvent {
    target: string; // CSS Selector
    name: string;
}

export interface FdTypeEvent extends FdEvent {
    target: string; // CSS Selector
    value: string;
}

export interface FdClickEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdHoverEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdLocationEvent extends FdEvent {
    href: string;
}

export interface FdLocationContainsEvent extends FdEvent {
    value: string;
}

export interface FdVisitEvent extends FdEvent {
    href: string;
}

export interface FdTextContentEvent extends FdEvent {
    target: string; // CSS Selector
    value: string;
}

export interface FdViewportSizeEvent extends FdEvent {
    width: number;
    height: number;
}

export interface FdExistsEvent extends FdEvent {
    target: string; // CSS Selector
}

export interface FdCountEqualsEvent extends FdEvent {
    target: string; // CSS Selector
    value: number;
}

export interface FdCountGreaterThanEvent extends FdEvent {
    target: string; // CSS Selector
    value: number;
}

export interface FdCountLessThanEvent extends FdEvent {
    target: string; // CSS Selector
    value: number;
}

export type AllFdEvents = FdEvent
    | FdAttributeValueEvent
    | FdAttributeExistsEvent
    | FdClickEvent
    | FdCountEqualsEvent
    | FdCountGreaterThanEvent
    | FdCountLessThanEvent
    | FdExistsEvent
    | FdHoverEvent
    | FdLocationEvent
    | FdLocationContainsEvent
    | FdTextContentEvent
    | FdTypeEvent
    | FdViewportSizeEvent
    | FdVisitEvent
;

export interface Template {
    name: string;
    description?: string;
    events: AllFdEvents[];
}

export const UNIQUE_SELECTOR_OPTIONS = {
    selectorTypes: ['Tag', 'NthChild']
};

export interface Options {
    basicAuth?: boolean;
    headers?: Header[];
}
