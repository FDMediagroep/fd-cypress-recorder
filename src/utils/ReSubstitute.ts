type KeyOrKeys = any | any[];

type Callback = (keys?: any[]) => void;

interface Subscription {
    id: number;
    callback: Callback;
    key: any;
}

interface SubscriptionOptions {
    keys?: any | any[];
    throttledUntil: number;
    bypassBlock: boolean;
}

/**
 * Drop-in minimal replacement for microsoft/ReSub.
 * Only implements the Pub/Sub functionality of ReSub.
 * Disregarded the framework specific implementations.
 */
export class ReSubstitute {
    static Key_All = 'RESUBSTITUTE_ALL_EVENTS';
    private subscriptions: Subscription[] = [];
    private static pendingCallbacks: Map<
        Callback,
        SubscriptionOptions
    > = new Map();
    private throttleMs = 0;
    private bypassTriggerBlocks = false;
    private static triggerBlockCount = 0;

    constructor(throttleMs = 0, bypassTriggerBlocks = false) {
        this.throttleMs = throttleMs;
        this.bypassTriggerBlocks = bypassTriggerBlocks;
    }

    private static resolveCallbacks() {
        for (const [callback, options] of ReSubstitute.pendingCallbacks) {
            if (ReSubstitute.triggerBlockCount && !options.bypassBlock) {
                // The callback does not bypass the block so we continue to the next callback.
                continue;
            } else if (options.bypassBlock) {
                callback.call(null, options.keys);
            } else if (+new Date() >= options.throttledUntil) {
                callback.call(null, options.keys);
            } else {
                setTimeout(
                    callback.bind(null, options.keys),
                    +new Date() - options.throttledUntil
                );
            }
            ReSubstitute.pendingCallbacks.delete(callback);
        }
    }

    /**
     * Subscribe to store triggers.
     * @param callback
     * @param key limit only to events for this key
     */
    subscribe(callback: Callback, key?: string): number {
        console.log('sub', key);
        const id = +new Date();
        this.subscriptions.push({ id, callback, key });
        return id;
    }

    /**
     * Unsubscribe from the store so future triggers will not trigger the callback anymore.
     * @param subToken
     */
    unsubscribe(subToken: number) {
        this.subscriptions = [
            ...this.subscriptions.filter(
                (subscription) => subscription.id !== subToken
            ),
        ];
    }

    protected _getSubscriptionKeys(): (string | undefined)[] {
        return this.subscriptions.map((subscription) => {
            return subscription.key;
        });
    }

    protected _isTrackingKey(key: string) {
        const results = this.subscriptions.find(
            (subscription) => subscription.key === key
        );
        return !!results;
    }

    /**
     * Trigger callbacks of subscriptions.
     * @param keyOrKeys trigger callback if subscription matches given key or keys.
     */
    trigger(keyOrKeys?: KeyOrKeys) {
        if (typeof keyOrKeys === 'string') {
            this.subscriptions.forEach((subscription) => {
                console.log(subscription.key ?? 'no key', keyOrKeys);
                if (
                    subscription.key === keyOrKeys ||
                    subscription.key === ReSubstitute.Key_All
                ) {
                    ReSubstitute.pendingCallbacks.set(subscription.callback, {
                        bypassBlock: this.bypassTriggerBlocks,
                        keys: [keyOrKeys],
                        throttledUntil: +new Date() + this.throttleMs,
                    });
                }
            });
        } else if (Array.isArray(keyOrKeys)) {
            this.subscriptions.forEach((subscription) => {
                console.log(subscription.key ?? 'no key', keyOrKeys);
                if (
                    keyOrKeys.indexOf(subscription.key) !== -1 ||
                    subscription.key === ReSubstitute.Key_All
                ) {
                    ReSubstitute.pendingCallbacks.set(subscription.callback, {
                        bypassBlock: this.bypassTriggerBlocks,
                        keys: keyOrKeys,
                        throttledUntil: +new Date() + this.throttleMs,
                    });
                }
            });
        } else {
            // Notify all listeners
            this.subscriptions.forEach((subscription) => {
                ReSubstitute.pendingCallbacks.set(subscription.callback, {
                    bypassBlock: this.bypassTriggerBlocks,
                    keys: keyOrKeys,
                    throttledUntil: +new Date() + this.throttleMs,
                });
            });
        }

        ReSubstitute.resolveCallbacks();
    }

    /**
     * Block callback triggers globally.
     */
    static pushTriggerBlock() {
        this.triggerBlockCount++;
    }

    /**
     * Pop trigger block. When no trigger blocks are left. The callback triggering will resume.
     * If no trigger blocks are set then this function does nothing.
     */
    static popTriggerBlock() {
        if (this.triggerBlockCount) {
            this.triggerBlockCount--;
            if (this.triggerBlockCount === 0) {
                ReSubstitute.resolveCallbacks();
            }
        }
    }
}
