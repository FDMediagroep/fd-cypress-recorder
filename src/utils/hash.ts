declare let Uint8Array: Uint8ArrayConstructor;
/**
 * Convert given ArrayBuffer to HEX-string.
 * @param buffer
 */
export function hexString(buffer: ArrayBuffer) {
    const byteArray = new Uint8Array(buffer);

    const hexCodes = [...byteArray].map((value) => {
        const hexCode = value.toString(16);
        const paddedHexCode = hexCode.padStart(2, '0');
        return paddedHexCode;
    });

    return hexCodes.join('');
}

/**
 * Digest the given message. ArrayBuffer result is returned in a Promise.
 * @param message
 */
export function digestMessage(message: string) {
    const encoder = new TextEncoder();
    const data = encoder.encode(message);
    return window.crypto.subtle.digest('SHA-256', data);
}

/**
 * Returns digest of given message in the form of a HEX-string.
 * @param message
 */
export async function digest(message: string) {
    const arrayBuffer = await digestMessage(message);
    const hex = hexString(arrayBuffer);
    return hex;
}

/**
 * Returns true when the sum of the ArrayBuffer digest of given message is even and false when odd.
 * @param message
 */
export async function isEven(message: string) {
    const arrayBuffer = await digestMessage(message);
    const byteArray = new Uint8Array(arrayBuffer);
    return (
        byteArray.reduce(
            (previousValue: number, currentValue: number) =>
                previousValue + currentValue
        ) %
            2 ===
        0
    );
}
