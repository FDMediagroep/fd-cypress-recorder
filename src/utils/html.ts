/**
 * Embed given string as JavaScript in the user-DOM space.
 * @param fn
 */
export function embed(fn: (text: string) => void) {
    const script = document.createElement("script");
    script.text = `(${fn.toString()})();`;
    document.documentElement.appendChild(script);
}

/**
 * Convert the given HTML-string to a DOM-node.
 * @param html
 */
export function htmlToElement(html: string): ChildNode {
    const template = document.createElement('template');
    html = html.trim(); // Never return a text node of whitespace as the result
    template.innerHTML = html;
    return template.content.firstChild as ChildNode;
}
