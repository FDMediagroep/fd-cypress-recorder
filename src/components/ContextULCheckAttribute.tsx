import React from "react";
import { FdEventType, AllFdEvents } from "../utils/CypressDictionary";
import styled from "styled-components";

export interface Props {
    target: HTMLElement;
    selector: string;
    onBack?: () => void;
    onMouseDown: (event: AllFdEvents) => void;
}

export default function ContextULCheckAttribute(props: Props) {
    function handleAttributeValueEquals(e: React.MouseEvent<HTMLElement>) {
        props.onMouseDown({
            type: FdEventType.ATTRIBUTE_VALUE_EQUALS,
            target: props.selector,
            name: e.currentTarget.getAttribute('data-name') as string,
            value: e.currentTarget.getAttribute('data-value') as string
        });
    }

    function handleAttributeValueContains(e: React.MouseEvent<HTMLElement>) {
        const value = prompt('Attribute value contains');
        if (value) {
            props.onMouseDown({
                type: FdEventType.ATTRIBUTE_VALUE_CONTAINS,
                target: props.selector,
                name: e.currentTarget.getAttribute('data-name') as string,
                value
            });
        }
    }

    function handleAttributeExists(e: React.MouseEvent<HTMLElement>) {
        props.onMouseDown({
            type: FdEventType.ATTRIBUTE_VALUE_EXISTS,
            target: props.selector,
            name: e.currentTarget.getAttribute('data-name') as string
        });
    }

    function handleBack(e: React.MouseEvent<HTMLElement>) {
        e.preventDefault();
        if (props.onBack) { props.onBack(); }
    }

    return (
        <ul>
            {(props.target as HTMLElement).attributes.length === 0 ? <li className="label back" onMouseDown={handleBack}>&lt; No attributes found</li> : <li className="label back"onMouseDown={handleBack}>&lt; Attributes</li>}
            {
                [].slice.call((props.target as HTMLElement).attributes).map((attribute: any, idx: number) => {
                    if (typeof attribute === 'object') {
                        return (
                            <React.Fragment key={attribute.name}>
                                <li className="label separator"><b>{attribute.name}</b></li>
                                <li className="clickable" data-name={attribute.name} data-value={attribute.value} onMouseDown={handleAttributeValueContains}>Contains...</li>
                                <li className="clickable" data-name={attribute.name} data-value={attribute.value} onMouseDown={handleAttributeValueEquals}><StyledDiv>Equals<small>{attribute.value}</small></StyledDiv></li>
                                <li className="clickable" data-name={attribute.name} data-value={attribute.value} onMouseDown={handleAttributeExists}>Exists</li>
                            </React.Fragment>
                        );
                    }
                    return null;
                })
            }
        </ul>
    );
}

const StyledDiv = styled.div`
    display: flex;
    flex-direction: column;
    small {
        padding: 0;
    }
`;
