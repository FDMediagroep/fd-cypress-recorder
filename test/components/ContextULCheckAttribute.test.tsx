import React from 'react';
import TestRenderer, { act } from 'react-test-renderer';
import ContextULCheckAttribute from '../../src/components/ContextULCheckAttribute';
import ReactDOM from 'react-dom';
import { FdEventType } from '../../src/utils/CypressDictionary';

describe('Context Menu Attribute', () => {
    let target: HTMLElement;
    const existsMock = jest.fn();
    const backMock = jest.fn();
    const mouseDownEvt = new MouseEvent('mousedown', {bubbles: true});

    beforeEach(() => {
        target = document.createElement('div');
        target.classList.add('context-menu-attributes-test');
        document.body.appendChild(target);
        act(() => {
            ReactDOM.render(<ContextULCheckAttribute onBack={backMock} onMouseDown={existsMock} selector=".context-menu-attributes-test" target={target}/>, target);
        });
    });

    afterEach(() => {
        document.body.removeChild(target);
        target.remove();
        existsMock.mockClear();
        backMock.mockClear();
    });

    it('should render without attributes correctly', () => {
        let contextMenu = TestRenderer.create(<ContextULCheckAttribute onMouseDown={() => {}} selector="document.body" target={document.body}/>);
        expect(contextMenu.toJSON()).toMatchSnapshot();
    });

    it('should render with attributes correctly', () => {
        let contextMenu = TestRenderer.create(<ContextULCheckAttribute onMouseDown={() => {}} selector=".context-menu-attributes-test" target={target}/>);
        expect(contextMenu.toJSON()).toMatchSnapshot();
    });

    it('should handle Exists event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase() === 'exists') {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(existsMock).toHaveBeenCalledTimes(1);
        expect(existsMock).toBeCalledWith({
            type: FdEventType.ATTRIBUTE_VALUE_EXISTS,
            target: '.context-menu-attributes-test',
            name: 'class'
        });
    });

    it('should handle Equals event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.textContent.toLowerCase().indexOf('equals') > -1) {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(existsMock).toHaveBeenCalledTimes(1);
        expect(existsMock).toBeCalledWith({
            type: FdEventType.ATTRIBUTE_VALUE_EQUALS,
            target: '.context-menu-attributes-test',
            name: 'class',
            value: 'context-menu-attributes-test'
        });
    });

    it('should handle Back event correctly', () => {
        [].slice.call(target.querySelectorAll('li')).forEach((li: HTMLLIElement) => {
            if (li.textContent && li.classList.contains('back')) {
                li.dispatchEvent(mouseDownEvt);
            }
        });
        expect(backMock).toHaveBeenCalledTimes(1);
    });

});
