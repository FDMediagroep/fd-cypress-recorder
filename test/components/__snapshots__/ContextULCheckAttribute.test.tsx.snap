// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Menu Attribute should render with attributes correctly 1`] = `
<ul>
  <li
    className="label back"
    onMouseDown={[Function]}
  >
    &lt; Attributes
  </li>
  <li
    className="label separator"
  >
    <b>
      class
    </b>
  </li>
  <li
    className="clickable"
    data-name="class"
    data-value="context-menu-attributes-test"
    onMouseDown={[Function]}
  >
    Contains...
  </li>
  <li
    className="clickable"
    data-name="class"
    data-value="context-menu-attributes-test"
    onMouseDown={[Function]}
  >
    <div
      className="sc-bdVaJa bRGwQJ"
    >
      Equals
      <small>
        context-menu-attributes-test
      </small>
    </div>
  </li>
  <li
    className="clickable"
    data-name="class"
    data-value="context-menu-attributes-test"
    onMouseDown={[Function]}
  >
    Exists
  </li>
</ul>
`;

exports[`Context Menu Attribute should render without attributes correctly 1`] = `
<ul>
  <li
    className="label back"
    onMouseDown={[Function]}
  >
    &lt; No attributes found
  </li>
</ul>
`;
